lblod/ember-rdfa-editor-lblod-plugins
# add `template comments` support

### Overview
This adds "template comments": a block of text that can be inserted "anywhere" to give some extra explanation how to fill in a document.
This text is only informational and should be hidden at publishing (_currently not implemented yet_).

Also bumps editor to [4.2.0](https://github.com/lblod/ember-rdfa-editor/releases/tag/v4.2.0). This does not include any breaking changes, but does include changes needed for this PR.

##### connected issues and PRs:
- [template comments GN-4264](https://binnenland.atlassian.net/browse/GN-4264)
- [ember-rdfa-editor#908 PR for embedded-editor changes](https://github.com/lblod/ember-rdfa-editor/pull/908)

### Setup
Need https://github.com/lblod/ember-rdfa-editor/pull/908
- use `npm link` in `ember-rdfa-editor`
- use `npm link @lblod/ember-rdfa-editor` in this repo
- run `ember ts:precompile' in `ember-rdfa-editor` to compile the types and avoid those giving errors.

OR, now that [ember-rdfa-editor 4.2.0](https://github.com/lblod/ember-rdfa-editor/releases/tag/v4.2.0) is released:
- no setup needed - 
### How to test/reproduce
- Go to Regulatory Statements "tab"
- insert a template comment with the button under "insert"
- you can type in the template comment and do everything specified in the issue:

see [comment on design](https://binnenland.atlassian.net/browse/GN-4277?focusedCommentId=79273) what should specifically be implemented: 
- indenting
- numbering
- bullet points
- bold text formatting 

On the right side buttons appear to move and delete the template comment.

### Challenges/uncertainties
~~Some known bugs. Specific of embedded editor are written in [#908 PR](https://github.com/lblod/ember-rdfa-editor/pull/908)~~ => these were a problem with the configuration of the schema in the plugins repo specifically. This has been fixed.

Other problems:

<details>
<summary>Details</summary>

- The list plugin has problems.
  1. create a template comment
  2. type some text
  3. try to make it a list item
  4. an error will occur that an undefined value got unwrapped. This is in the `dispatchInner` of the embedded-editor
  Note that it does work if you copy a list item directly as the first item. It seems like the actual wrapping logic of pressing the list-item button is the problem.

- ~~undo seems to not work correctly anymore (also a problem with transactions?)~~ For some reason [this commit](https://github.com/lblod/ember-rdfa-editor/pull/908/commits/008dcd53b49305271004b0e197ff918b20813eed) fixes the problem
- Italicalizing is not needed. The button (CTRL+I) is disabled, but pressing the button still adds it.
  - Easiest solution is probably: add a "non-italic-paragraph" node to use as content for the template comment

</details>

Problems have been fixed by changing from embedded-editor (which should only be used for inline content) to just using content logic of ember node.
- lists now work as expected
- italize has been "disabled" by using a paragraphWithConfig

#### Move buttons
These have been implemented in an easy way to just move between nodes, as it wasn't really specified in the design and "moving up or down" in general is kind of a hard thing to specify (and probably not desired for a comment?).
The implementation seems to not work for nested nodes because `nodeBefore` and `nodeAfter` don't return anything. A better implementation might be to use nodeBefore if possible, and otherwise change to findInAncestor (like the structures-plugin) to find a fitting (allowed) node to insert.
